### PR TITLE
chore(docs): add PR manifesto

### PR DESCRIPTION
**Pull request by guoda-puidokaite**
# Summary

Added a PR Manifesto to the Juno repository.

**Action for reviewers:** This is an initial proposal. PLEASE LEAVE FEEDBACK. 😊 Or reach out directly to discuss.

# Changes Made

- Documented 10 PR best practise points, addressing concerns and suggestions raised in our Retros.
- Added a link to the PR template

# Related Issues

- https://github.com/cloudoperators/juno/issues/642

# Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have made corresponding changes to the documentation (if applicable).
- [ ] My changes generate no new warnings or errors.
- [ ] I have created a changeset for my changes.

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -38,3 +38,7 @@
 - [ ] I have made corresponding changes to the documentation (if applicable).
 - [ ] My changes generate no new warnings or errors.
 - [ ] I have created a changeset for my changes.
+
+# PR Manifesto
+
+Review the [PR Manifesto](https://github.com/cloudoperators/juno/blob/main/docs/pr_manifesto.md) for best practises.
```
Comment (guoda-puidokaite) [inline .github/pull_request_template.md:44]: Where would be the best place to add the PR Manifesto? The checklist is mainly for the PR owner, so though we should mention it outside? But down mind doing it within.

File: docs/pr_manifesto.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+## 3. Prioritise Code Quality
+
+- Develop code that is readable, efficient, and maintainable.
+- Adhere to coding best practices such as clear naming conventions, modular design, and thorough documentation for complex logic.
+- Follow established repository conventions and standards to ensure consistency.
+
+## 4. Test Thoroughly
+
+- Implement robust tests to verify functionality and support continuous maintenance.
+- Share responsibility for quality assurance, ensuring that no regressions are introduced by new logic.
```
Comment (edda) [inline docs/pr_manifesto.md:25]: Suggest to add the bit about adding manual testing instructions
```suggestion
- Share responsibility for quality assurance, ensuring that no regressions are introduced by new logic.
- Add instructions for manual testing if they are not obvious (e.g. if a specific edge case must be tested)
- Manually test app at runtime where possible
```

Comment (guoda-puidokaite) [reply docs/pr_manifesto.md]: Thank You for the fast feedback, @edda! 

Manual Testing: The original comment on quality assurance was intended to include this. I made it more explicit now, Thank You for the suggestion.

Testing Instructions: I included this in Point 2, it's a good additional to the documentation part of the PR.

Regarding QA, I'd be happy to further discuss improvements we could make and contribute to documentation of a separate strategy too.

**Review by edda — APPROVED** [review]

File: docs/pr_manifesto.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+- All contributors are responsible for the prevention of regressions by thoroughly testing new functionality, including manual runtime verification.
+- When transferring an identified issue to a new ticket, investigate the root cause, confirm it’s outside the PR’s scope, and link the tickets for traceability.
+
+## 5. Give Actionable and Prioritised Feedback
+
+- Offer specific and constructive feedback using the "Praise, Question, Suggest" structure:
+  - **Praise** positive aspects.
+  - **Question** to gain understanding.
+  - **Suggest** improvements with clear technical reasoning and examples.
+- "Request changes" only for critical issues affecting functionality or code quality, and label other feedback as a "Question" or "Non-Blocking Improvement".
```
Comment (ArtieReus) [inline docs/pr_manifesto.md:35]: I really like this sentence, and I would emphasize it even more.

File: docs/pr_manifesto.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+## 6. Take Ownership
+
+- Own your code, be prepared to explain your decisions, and be open to suggestions, recognizing the value of diverse perspectives in achieving best outcomes.
+
+## 7. Address Comments
+
+- Understand each PR comment and take one of three actions:
```
Comment (ArtieReus) [inline docs/pr_manifesto.md:43]: Wouldn't be also an option:
- Accept the feedback but defer the change to a separate PR, explaining the reason.

**Review by ArtieReus — COMMENTED** [review]
This sounds to me very good. I just add a comment.

File: docs/pr_manifesto.md (hunk 3 above)
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md:35]: Thank You. How would you like to emphasise it?

File: docs/pr_manifesto.md (hunk 4 above)
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md:43]: Great suggestion, Thank You. I've updated the document with instructions for deferring a change, please have a look. 🙏

File: .github/pull_request_template.md (hunk 1 above)
Comment (ArtieReus) [reply .github/pull_request_template.md:44]: docs is probably the right place I would say..

**Review by ArtieReus — APPROVED** [review]
lgtm

File: .github/pull_request_template.md (hunk 1 above)
Comment (guoda-puidokaite) [reply .github/pull_request_template.md:44]: Ah, sorry! I updated my comment... I meant to ask if anyone has another suggestion of where the _link_ to the PR Manifesto should be added. 🙈 But so far, looks like the current location is appropriate.

File: docs/pr_manifesto.md (hunk 3 above)
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md]: I made some improvements, please resolve if all good. 🙏

**Comment by franzheidl** [comment]
Very nice! 👍

**Review by TilmanHaupt — COMMENTED** [review]
i think its good - added two questions.

File: docs/pr_manifesto.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+- Ensure all changes strictly adhere to the ticket’s acceptance criteria.
+- Any deviations must be agreed upon with the Product Owner and documented as updates in the original ticket.
+
+## 2. Summarise Changes
+
+- Provide a comprehensive summary of changes and key technical decisions:
+  - Highlight performance impacts, dependencies, and links to relevant documentation.
+  - Use visuals to enhance clarity.
+  - Include clear instructions for manual testing of complex or non-obvious edge cases.
```
[the remark this answers is not in the document]
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md:15]: Thank You for the review! 

From our current agreement in the team, every contributor of a PR is responsible for throughly testing it (as in point 4):
- **Developers** should consider all happy and edge cases during development and ensure that everything works correctly before submitting for review.
- **Reviewers** are the gatekeepers who test that the criteria are fulfilled and that everything works as expected, since the developer is often too close to the code to remain objective. Even if a developer had a case in mind while developing and testing, it should be tested again. This point is a reminder that it's helpful if the developer leaves tips on complicated cases to ease this process.

However, there's many aspects to discuss here and ways we could improve the process. See more background [here](https://github.com/cloudoperators/juno/pull/784#discussion_r1960205727).

File: docs/pr_manifesto.md (hunk 2 above)
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md]: Linking a discussion on this point: https://github.com/cloudoperators/juno/pull/784#discussion_r1965131959

File: docs/pr_manifesto.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md:15]: Does that address your question, wdyt?

File: docs/pr_manifesto.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+- Keep PRs concise, well-organized, and easy to understand to enable swift reviews and merging. Remember that splitting functionality into smaller PRs can improve clarity and manageability.
```
Comment (TilmanHaupt) [inline docs/pr_manifesto.md:16]: I am now against this approach - because it makes thinks less effient. Big PR are maybe redone or thrown away but right now - after i split my PR i really hate it: You code like 2 days and then your blocked because you need 4 days to approve it. While you are on this road the PR gets more and more out of scope and you just do a lot of stuff from later tickets - because of discussions and so on.

File: docs/pr_manifesto.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+- Implement robust tests to verify functionality and support continuous maintenance.
+- All contributors are responsible for the prevention of regressions by thoroughly testing new functionality, including manual runtime verification.
+- When transferring an identified issue to a new ticket, investigate the root cause, confirm it’s outside the PR’s scope, and link the tickets for traceability.
+
+## 5. Give Actionable and Prioritised Feedback
+
+- Contribute to ongoing discussions to ensure focused communication and prevent duplication on related topics.
+- Offer specific and constructive feedback using the "Praise, Question, Suggest" structure:
+  - **Praise** positive aspects.
```
[the remark this answers is not in the document]
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md:35]: I see your concern. 🙈 

No one is forced to give praise or say anything they don't mean. This structure is just a reminder of good PR and team culture. 

While pointing out areas of improvement is crucial, looking at a PR through a positive lens can be just as valuable. Genuine positive feedback provides a balanced perspective, recognises good work, and tells a developer what they should keep doing.

Let me know if the wording can be improved or if you still have concerns.

File: docs/pr_manifesto.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (TilmanHaupt) [reply docs/pr_manifesto.md:15]: Partially: "Even if a developer had a case in mind while developing and testing, it should be tested again." - I agree on that. I just meant that if the tester does 100% the same like the dev the same will come out or? So its good if the tester test the own ways and cases for the same goal?  But I guess the part is ok - but i just thought its maybe lead to the same blindness for the tester - because he tested like the dev would. And not testing the feature how a user would.

File: docs/pr_manifesto.md (hunk 6 above)
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md:16]: Sorry to hear about those experiences. It sounds like there were several challenges that need to be untangled, and therefore, it's a very valuable topic for our next Retro - please bring it up.

However, there's no clear approach outlined here to be against. It's very subjective what a small or big PR is. It's also very dependent on the ticket and what functionality can or can't be split efficiently. For instance, it's not a good idea to split highly coupled functionality.

The only recommendation here is to think of the reviewer and not just the delivery of functionality. Your PR should be concise, organized, and easy to understand. Indeed, splitting it up can improve clarity and manageability. But how you approach that—what functionality you split or don't split, and to some degree, the efficiency—is up to you.

Do you have any specific suggestions on how the guidance can be improved?

Comment (TilmanHaupt) [reply docs/pr_manifesto.md:16]: With "big" i just meant delivering the hole feature and not a single ticket / sub feature. (like store slice)

File: docs/pr_manifesto.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md:15]: I totally agree with you! It's what I'd like to discuss when we're developing our QA strategy. 

Firstly, I think that the test cases for a ticket should be determined during refinement and live in the ticket instead of the PR... since the ticket is the source of truth for requirements. If new ones are found, they can be added to the ticket as updates. 

Secondly, regarding objectivity: Totally agree. Even if the developer left some steps on what they found was important, testing should remain objective and they shouldn't be relied upon too much. The testers should perform their own exploratory testing, testing like a user.

@edda, what do you think about this? What did you have in mind when suggesting to add this point?

[the remark this answers is not in the document]
Comment (edda) [reply docs/pr_manifesto.md:15]: I agree with the point you raised @TilmanHaupt regarding essentially mentally boxing in the reviewer by describing all the steps in details. That's also not what I meant. Rather than writing detailed steps how to test we should describe **what** to test (in terms of user story) and leave the reviewer to find their own path there. Since the wording now causes confusing let's try to improve it or remove this point for now until we can come up with something better.

[the remark this answers is not in the document]
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md:15]: I'll remove it until we can discuss our wider quality approach.

File: docs/pr_manifesto.md (hunk 6 above)
Comment (barsukov) [reply docs/pr_manifesto.md:16]: I think the goal here is to reduce the cognitive load, so means do not do more then it expected.
good rule of thumb One job = One PR with maybe max deviation of 10-20%.
As example: 
So if you change domain "Regions" and then you starting updating Domain "Service" paralel upgrading react-query api calls everywhere you should ask yourself that you are sligtly off grid maybe :)

Comment (barsukov) [reply docs/pr_manifesto.md:16]: > With "big" i just meant delivering the hole feature and not a single ticket / sub feature. (like a store slice with a few helpers)

I could be slightly disagree with you on this one with my typical answer "It depends" sometimes it is better to split your big feature on well readable and testable chunks then deliver a big bang thing.
So again it hard to define, cause feature weight could be really diferent. 
Again main point, when you deliver something it is not only "O, I am done with my job" it is also about the reviewer, you should always keep in mind the reviewer of your code, and try to reduce the cognitive load of your team members.

Comment (TilmanHaupt) [reply docs/pr_manifesto.md:16]: But you need to touch other files and people will not see your scope and block the PR. Also maybe for good reason, because its complicated to know if you come back.

File: docs/pr_manifesto.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md]: I rephrased slightly - have a look. 😉

**Review by barsukov — APPROVED** [review]

**Review by TilmanHaupt — APPROVED** [review]

File: docs/pr_manifesto.md (hunk 6 above)
Comment (guoda-puidokaite) [reply docs/pr_manifesto.md]: @TilmanHaupt @barsukov 

Closing this thread unless you can share specific suggestions on how the guidance can be improved?

**Review by taymoor89 — APPROVED** [review]
Sounds good, thank you for doing the effort to prepare this 👍

**Review by hgw77 — APPROVED** [review]

**Review by hodanoori — APPROVED** [review]

**Review by andypf — APPROVED** [review]

**Review by MartinS-git — APPROVED** [review]
